src/text-input/src/TextInput.js

```diff
@@ -89,25 +118,55 @@ class TextInput extends PureComponent {
     const borderRadius = theme.getBorderRadiusForControlHeight(height)
 
     return (
-      <Text
-        is="input"
-        className={cx(themedClassName, className)}
-        type="text"
-        size={textSize}
-        width={width}
-        height={height}
-        required={required}
-        disabled={disabled}
-        placeholder={placeholder}
-        paddingLeft={Math.round(height / 3.2)}
-        paddingRight={Math.round(height / 3.2)}
-        borderRadius={borderRadius}
-        spellCheck={spellCheck}
-        aria-invalid={isInvalid}
-        {...(disabled ? { color: 'muted' } : {})}
-        css={css}
-        {...props}
-      />
+      <React.Fragment>
+        <Text
+          is="input"
+          className={cx(themedClassName, className)}
+          type="text"
+          size={textSize}
+          width={width}
+          height={height}
+          required={required}
+          disabled={disabled}
+          placeholder={placeholder}
+          value={this.getValue()}
+          {...(onChange
+            ? { onChange }
+            : { onChange: e => this.setState({ value: e.target.value }) })}
+          paddingLeft={Math.round(height / 3.2)}
+          paddingRight={Math.round(height / 3.2)}
+          borderRadius={borderRadius}
+          spellCheck={spellCheck}
+          aria-invalid={isInvalid}
+          {...(disabled ? { color: 'muted' } : {})}
+          css={css}
+          {...props}
+        />
+        {isClearable &&
+          (!this.props.value &&
+            this.state.value.length > 0 &&
+            !this.props.onChange) && (
+            <Icon
+              color="muted"
+              icon="cross"
+              appearance="default"
+              onClick={() => {
+                this.setState({ value: '' })
+              }}
+              marginLeft={-20}
+            />
+          )}
+        {isClearable &&
+          (this.props.value && this.props.onChange) && (
```

Comment:
Is there a good reason to have two separate `isClearable` behaviors?

Reply:
Probably a naive reason way back when, but I think I was just confusing myself before. I think I know how to collapse now.